### PR TITLE
Implemented new AST changes

### PR DESCRIPTION
**Pull request by pksunkara**
Along with the source maps. Introduced a new `Element` structure. `ResourceGroup` structure is converted into a shim of `Element` so that `SectionProcessor<ResourceGroup>` works. Similarly, `DataStructureGroup`.

**NOTE**: This PR do not have serialisation or c-interface, thus disabling their tests.

### CODE REVIEW

File: src/Blueprint.h [hunk 1]
```diff
@@ -16,6 +16,9 @@
 #include "MarkdownNode.h"
 #include "MSONSourcemap.h"
 
+#define RESOURCE_NOT_SET_ERR std::logic_error("no resource set")
+#define DATA_STRUCTURE_NOT_SET_ERR std::logic_error("no data structure set")
+
```
Comment (pksunkara) [inline src/Blueprint.h]: Will remove these 3 lines.

**Comment by zdne** [comment]
@pksunkara this one is failing, please fix before the review

**Comment by pksunkara** [comment]
Updated. Forgot to remove the integration tests

File: .travis.yml [hunk 2]
```diff
@@ -3,7 +3,7 @@ compiler:
   - gcc
   - clang
 script: 
-  - ./configure --include-integration-tests
+  - ./configure
```
Comment (zdne) [inline .travis.yml]: we better not forget about this :8ball:

Comment (pksunkara) [inline .travis.yml]: Already added this back in #279

File: src/Blueprint.h [hunk 3]
```diff
@@ -349,23 +346,106 @@ namespace snowcrash {
...
+        Element& operator=(const Element& rhs);
+
+        /** Destructor */
+        ~Element();
+    };
+
+    /**
+     *  Group of API Resources (Category Element)
+     */
+    struct ResourceGroup : public Element {
```
Comment (zdne) [inline src/Blueprint.h]: Why not just typedef?

File: src/Blueprint.h [hunk 4]
```diff
@@ -349,23 +346,106 @@ namespace snowcrash {
...
     };
 
-    /** Collection of Resource groups */
-    typedef Collection<ResourceGroup>::type ResourceGroups;
+    /**
+     * Group of Data Structures (Category Element)
+     */
+    struct DataStructureGroup : public Element {
```
Comment (zdne) [inline src/Blueprint.h]: ditto

File: src/DataStructureGroupParser.h [hunk 5]
```diff
@@ -0,0 +1,154 @@
+//
+//  DataStructureGroupParser.h
```
Comment (zdne) [inline src/DataStructureGroupParser.h]: would be good you you could `git mv` the files next time when you are renaming, so I can actually see what has been changed.

File: src/ResourceGroupParser.h [hunk 6]
```diff
@@ -183,22 +237,63 @@ namespace snowcrash {
...
-                    return std::make_pair(it, match);
+            return true;
+        }
+
+        /**
+         * \brief Given list of elements, get the last resource
+         *
+         * \param elements Collection fo elements
+         */
+        static Resource lastResource(const Elements& elements) {
```
Comment (zdne) [inline src/ResourceGroupParser.h]: this is not the most elegant solution, you probably want to learn about [find_if](http://www.cplusplus.com/reference/algorithm/find_if/) & predicates

File: test/test-DataStructureGroupParser.cc [hunk 7]
```diff
@@ -0,0 +1,114 @@
+//
```
Comment (zdne) [inline test/test-DataStructureGroupParser.cc]: ditto, please use `git mv` next time

**Comment by zdne** [comment]
@pksunkara looks good please solve https://github.com/apiaryio/snowcrash/pull/278#discussion_r23696241 and its good to merge

File: src/DataStructureGroupParser.h (hunk 5 above)
Comment (pksunkara) [inline src/DataStructureGroupParser.h]: I have actually did `git mv`, but when the changes go above a certain %, git treats it as normal add and delete.

File: src/Blueprint.h (hunk 3 above)
Comment (pksunkara) [inline src/Blueprint.h]: Due to our architecture, just `typedef`ing them won't work. We need both `SectionProcessor<ResourceGroup>` and `SectionProcessor<DataStructureGroup>`, but using `typedef` will give redefinition errors.

File: src/ResourceGroupParser.h (hunk 6 above)
Comment (pksunkara) [inline src/ResourceGroupParser.h]: I tried searching for something which does reverse of find_if, but didn't find anything.

**Comment by pksunkara** [comment]
@zdne Addressed all your comments.

File: src/DataStructureGroupParser.h (hunk 5 above)
Comment (zdne) [inline src/DataStructureGroupParser.h]: oh thats not smart, OK didn't know that

File: src/ResourceGroupParser.h (hunk 6 above)
Comment (zdne) [inline src/ResourceGroupParser.h]: You need to provide your own predicate for matching elements based on element type. See other `std::find_if` use in snow crash.

**Pull request merged**